### PR TITLE
Backport: Reindex's throttling

### PR DESCRIPTION
**Pull request by nik9000**
Adds `requests_per_second` to `_update_by_query` and `_reindex`. Adds an action to dynamically rethrottle a running reindex request at `/_reindex/{taskId}/_rethrottle` and `/_update_by_query/{taskId}/_rethrottle`.

### CODE REVIEW

**Comment by nik9000** [comment]
@dakrone can you look at these? It is a relatively strait forward backport of code I believe you already reviewed.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 1]
```diff
@@ -944,6 +944,10 @@ public XContentBuilder dateValueField(XContentBuilderString rawFieldName, XConte
         return this;
     }
 
+    public XContentBuilder timeValueField(String rawFieldName, String readableFieldName, TimeValue timeValue) throws IOException {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: This seems weird, because if someone does:

```java
timeValueField("took_in_nanos", "took", TimeValue.timeValueNanos(1234L))
```

They'll get a field called "took_in_nanos" that's actually milliseconds!

File: core/src/test/java/org/elasticsearch/test/ESTestCase.java [hunk 2]
```diff
@@ -287,6 +287,34 @@ public static double randomDouble() {
...
+     * @param start          lower bound of interval to draw uniformly distributed random numbers from
+     * @param end            upper bound
+     * @param lowerInclusive whether or not to include lower end of the interval
+     */
+    public static double randomDoubleBetween(double start, double end, boolean lowerInclusive) {
+        double result = 0.0;
+
+        if (start == -Double.MAX_VALUE || end == Double.MAX_VALUE) {
+            // formula below does not work with very large doubles
+            result = Double.longBitsToDouble(randomLong());
```
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/test/ESTestCase.java]: Doing these with `randomLong()` seems kind of complex, is there a reason you can't do:

```java
double result = randomDouble();
while (result < start || result > end) {
  result = randomDouble();
}
return result;
```

?

File: core/src/test/java/org/elasticsearch/test/ESTestCase.java [hunk 3]
```diff
@@ -357,9 +385,16 @@ public static String randomRealisticUnicodeOfCodepointLength(int codePoints) {
         return array;
     }
 
+    public static String randomTimeValue(int lower, int upper) {
```
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/test/ESTestCase.java]: Not sure what worth having a lower and upper limit is if it's randomly seconds versus days?

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/test/ESTestCase.java]: This is what we have in master - I presume the lower limit is there so you can spit out a non 0 time value.

File: core/src/test/java/org/elasticsearch/test/ESTestCase.java (hunk 2 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/test/ESTestCase.java]: This also just comes straight from master. I didn't even read it because it is what we're using there....

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/test/ESTestCase.java]: I believe they do this because randomDouble gets you something between 0 and 1 but this gets you something all over the place.

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: I'll have a look. This might be another one of those "we should fix it it master" things. Or not.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java [hunk 4]
```diff
@@ -309,6 +335,21 @@ public void onFailure(Throwable e) {
...
+    float perfectlyThrottledBatchTime(int lastBatchSize) {
+        if (task.getRequestsPerSecond() == 0) {
+            return 0;
+        }
+        //       requests
+        // ------------------- == seconds
+        // request per seconds
+        float targetBatchTimeInSeconds = lastBatchSize / task.getRequestsPerSecond();
+        // nanoseconds per seconds * seconds == nanoseconds
+        return TimeUnit.SECONDS.toNanos(1) * targetBatchTimeInSeconds;
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java]: Could always make this a class var so it doesn't have to recalculate the number of nanoseconds in a second every time (but I guess it will be JIT-ed anyway)

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: I can just backport what is in master now for this, but it isn't much better:
```java
    public XContentBuilder timeValueField(String rawFieldName, String readableFieldName, TimeValue timeValue) throws IOException {
        if (humanReadable) {
            field(readableFieldName, timeValue.toString());
        }
        field(rawFieldName, timeValue.millis());
        return this;
    }
```

I believe you'll still end up with the same issue.

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: Maybe is it as simple as changing `rawFieldName` to `millisFieldName` and documenting the method properly?

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: Yeah, totally outside the scope of this issue, just something I notice. Documenting sounds good to me.

File: core/src/test/java/org/elasticsearch/test/ESTestCase.java (hunk 2 above)
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/test/ESTestCase.java]: Ahh, makes sense

File: core/src/test/java/org/elasticsearch/test/ESTestCase.java (hunk 3 above)
Comment (dakrone) [inline core/src/test/java/org/elasticsearch/test/ESTestCase.java]: Okay, still seems strange though :)

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java [hunk 5]
```diff
@@ -296,6 +318,7 @@ public void readFrom(StreamInput in) throws IOException {
         consistency = WriteConsistencyLevel.fromId(in.readByte());
         retryBackoffInitialTime = TimeValue.readTimeValue(in);
         maxRetries = in.readVInt();
+        requestsPerSecond = in.readFloat();
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java]: Since this is a backport, this needs version checking

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java [hunk 6]
```diff
@@ -309,6 +332,7 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeByte(consistency.id());
         retryBackoffInitialTime.writeTo(out);
         out.writeVInt(maxRetries);
+        out.writeFloat(requestsPerSecond);
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java]: Same here for version checking

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java [hunk 7]
```diff
@@ -287,4 +365,115 @@ void countVersionConflict() {
...
+                public void onFailure(Throwable t) {
+                    prepareBulkRequestRunnable.onFailure(t);
+                }
+            };
+            delayedPrepareBulkRequestReference.set(new DelayedPrepareBulkRequest(threadPool, getRequestsPerSecond(), delay, oneTime));
+        }
+    }
+
+    private void setRequestsPerSecond(float requestsPerSecond) {
+        if (requestsPerSecond == -1) {
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: This should probably be `< 0` in case someone passes -2 as the requests per second

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java [hunk 8]
```diff
@@ -99,7 +146,10 @@ public Status(StreamInput in) throws IOException {
             versionConflicts = in.readVLong();
             noops = in.readVLong();
             retries = in.readVLong();
+            throttled = TimeValue.readTimeValue(in);
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: These need more version guards

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java [hunk 9]
```diff
@@ -112,7 +162,10 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeVLong(versionConflicts);
             out.writeVLong(noops);
             out.writeVLong(retries);
+            throttled.writeTo(out);
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: And here

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java (hunk 7 above)
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: So it also looks like explicitly setting the RPS does this, but passing it in the constructor doesn't, maybe the constructor should use this logic as well?

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java [hunk 10]
```diff
@@ -287,4 +365,115 @@ void countVersionConflict() {
...
+        synchronized (delayedPrepareBulkRequestReference) {
+            setRequestsPerSecond(newRequestsPerSecond);
+
+            DelayedPrepareBulkRequest delayedPrepareBulkRequest = this.delayedPrepareBulkRequestReference.get();
+            if (delayedPrepareBulkRequest == null) {
+                // No request has been queued yet so nothing to reschedule.
+                return;
+            }
+
+            this.delayedPrepareBulkRequestReference.set(delayedPrepareBulkRequest.rethrottle(newRequestsPerSecond));
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: We don't need to worry about the get-then-set concurrency issue here right?

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * A request to change throttling on a task.
+ */
+public class RethrottleRequest extends BaseTasksRequest<RethrottleRequest> {
+    /**
+     * The throttle to apply to all matching requests in sub-requests per second. 0 means set no throttle and that is the default.
+     * Throttling is done between batches, as we start the next scroll requests. That way we can increase the scroll's timeout to make sure
+     * that it contains any time that we might wait.
+     */
+    private float requestsPerSecond = 0;
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java]: How do you feel about not setting this and forcing a user of the Java API to be explicit about the rethrottling amount? Otherwise someone can accidentally un-throttle because they forgot to call `.setRequestsPerSecond`

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportRethrottleAction.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+public class TransportRethrottleAction extends TransportTasksAction<BulkByScrollTask, RethrottleRequest, ListTasksResponse, TaskInfo> {
+    @Inject
+    public TransportRethrottleAction(Settings settings, ClusterName clusterName, ThreadPool threadPool, ClusterService clusterService,
+            TransportService transportService, ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(settings, RethrottleAction.NAME, clusterName, threadPool, clusterService, transportService, actionFilters,
+                indexNameExpressionResolver, new Callable<RethrottleRequest>() {
```
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportRethrottleAction.java]: I feel like we used to have a different standard other than creating a `Callable` for every time we passed these in, I just can't remember what it was...

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java (hunk 5 above)
Comment (nik9000) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java]: Man, it has been too long since I had to pay attention to that. I'll get it!

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/AsyncBulkByScrollActionTests.java [hunk 13]
```diff
@@ -357,6 +381,61 @@ public void testBulkRejectionsRetryAndFailAnyway() throws Exception {
...
+        assertThat(client.lastScroll.get().request.scroll().keepAlive().seconds(), either(equalTo(110L)).or(equalTo(109L)));
+
+        // Now we can simulate a response and check the delay that we used for the task
+        InternalSearchHit hit = new InternalSearchHit(0, "id", new Text("type"), Collections.<String, SearchHitField>emptyMap());
+        InternalSearchHits hits = new InternalSearchHits(new InternalSearchHit[] { hit }, 0, 0);
+        InternalSearchResponse searchResponse = new InternalSearchResponse(hits, null, null, null, false, false);
+        client.lastScroll.get().listener.onResponse(new SearchResponse(searchResponse, scrollId(), 5, 4, randomLong(), null));
+
+        // The delay is still 100ish seconds because there hasn't been much time between when we requested the bulk and when we got it.
+        assertThat(capturedDelay.get().seconds(), either(equalTo(100L)).or(equalTo(99L)));
```
Comment (dakrone) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/AsyncBulkByScrollActionTests.java]: Hope this doesn't bite us for really slow (read: Windows) CI servers...

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java (hunk 10 above)
Comment (nik9000) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: I believe that to be safe because we're doing it within a synchronized block. And I used a synchronized block because this is low contention and I wanted there to be less of a chance that we'd queue two copies of the request.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportRethrottleAction.java (hunk 12 above)
Comment (nik9000) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportRethrottleAction.java]: I'm not sure it matters that much because master just uses a constructor reference. The task management code looks like this but I expect other code uses other things. Maybe a `public static final Callable` declared in the class being built? `Script` used to have one of those.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java (hunk 11 above)
Comment (nik9000) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java]: That is how it works on the REST layer - if the field isn't on the request it gives the user a BAD_REQUEST. I can make this a `Float` and validate that it is set. Maybe I should file an issue and implement it in master and backport? That backport would be clean so it wouldn't need another review. Or I can fix it in master and then add the backport into this.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java (hunk 7 above)
Comment (nik9000) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: The constructor just calls this method. I can make it a `< 0` thing.

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/AsyncBulkByScrollActionTests.java (hunk 13 above)
Comment (nik9000) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/AsyncBulkByScrollActionTests.java]: It hasn't yet and I think it is pretty safe - the time is between the `setLastBatchStartTime` and `startNextScroll` and those values are in seconds. I think floating point rounding errors are larger. But I could factor the time check in startNextScroll in such a way that the test can spoof it and we don't need any time.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java (hunk 10 above)
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/BulkByScrollTask.java]: Okay, sounds good.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java (hunk 11 above)
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java]: I think it might be nice to have it behave the same for the Java API, where it needs to be explicitly set, but as you say, this can be a separate PR

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportRethrottleAction.java (hunk 12 above)
Comment (dakrone) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/TransportRethrottleAction.java]: Okay, no problem with this then

File: modules/reindex/src/test/java/org/elasticsearch/index/reindex/AsyncBulkByScrollActionTests.java (hunk 13 above)
Comment (dakrone) [inline modules/reindex/src/test/java/org/elasticsearch/index/reindex/AsyncBulkByScrollActionTests.java]: Nah, if it hasn't hit yet we'll handle it if it does in the future, don't worry about it.

**Comment by dakrone** [comment]
Left minor comments and questions, mostly needs the version guards for the backport.

Also, not sure how hard it would be, but it would be great to add a test for this with a mixed-version cluster and make sure nothing breaks

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java (hunk 11 above)
Comment (nik9000) [inline modules/reindex/src/main/java/org/elasticsearch/index/reindex/RethrottleRequest.java]: https://github.com/elastic/elasticsearch/pull/18028

**Comment by nik9000** [comment]
I'll have a look at doing a mixed version test in the morning. After fixing the serialization I expect it to work  in a mixed version cluster if the coordinating node is 2.4.

**Comment by nik9000** [comment]
Stalling until #18028 is merged.

**Comment by nik9000** [comment]
@dakrone I've added the POSITIVE_INFINITY stuff to this backport!

File: docs/reference/docs/reindex.asciidoc [hunk 14]
```diff
@@ -376,24 +397,24 @@ The responses looks like:
...
-            "created" : 0,
-            "deleted" : 0,
-            "batches" : 36,
-            "version_conflicts" : 0,
-            "noops" : 0
-          },
-          "description" : ""
-        }
-      }
+      "tasks" : [ {
```
Comment (dakrone) [inline docs/reference/docs/reindex.asciidoc]: Something I just noticed, does this break the format compatibility by instead of return an object using the task id as the key, it returns an array of tasks? If so, we need to mark this as breaking! (and put notes in the migration guide)

Comment (nik9000) [inline docs/reference/docs/reindex.asciidoc]: This was mostly just fixing out of date documentation. This backport adds new fields to the `status` but I think that is ok?

File: .gitignore [hunk 15]
```diff
@@ -37,6 +37,5 @@ html_docs
 
 # random old stuff that we should look at the necessity of...
 /tmp/
-backwards/
```
Comment (nik9000) [inline .gitignore]: This change is a mistake. I'll remove it.

**Comment by nik9000** [comment]
@dakrone, I hate to pull your attention back to 2.x stuff, but do you think this is ok? I've got to remove the change to `.gitignore` from this but otherwise? I'd like to put it to bed.

**Comment by dakrone** [comment]
LGTM (with the `.gitignore` thing too)

Thanks for working through this and adding lots of tests!

**Comment by nik9000** [comment]
@dakrone Thanks for reviewing and prompting lots of great changes to reindex! I'm excited to get useful for 2.x folks!
